scientist-softserv/atla-hyku
# i22-customize-generic-worktype

# Story
As a user I can create a Generic Work Type customized to this [spreadsheet](https://docs.google.com/spreadsheets/d/1cyTtAofi2g1sx0ATJuP7ws-77a7OzBXF/edit#gid=1372594593)

# Expected Behavior Before Changes
A basic Generic Work model is available.

# Expected Behavior After Changes
The Generic Work Type Customizations:
- 2 custom fields: institution/format
- Updates required fields & primary_terms
- Updates multiple true/false to appropriate fields on form
- Updates multiple true/false on appropriate views
- Updates facets
- Updates to have appropriate fields on the index/show pages (search results page/show page)
- Adds help text to locales (& languages)
- Adds controlled vocabulary for ATLA resource_type/institution/format (moved over from atla hyrax)
- Adds 300 character maximum to the index/show page description field (moved over from atla hyrax)
- Adds test coverage and updates pre-existing generic_work models in tests with updated configuration.

Ref Issues Inlcuded:
- https://github.com/scientist-softserv/atla-hyku/issues/27
- https://github.com/scientist-softserv/atla-hyku/issues/30
- https://github.com/scientist-softserv/atla-hyku/issues/43

# Screenshots / Video
Video creating and then editing a Generic Work: https://share.getcloudapp.com/04u6wOYQ
<details>
<summary>CLICK FOR SCREENSHOTS</summary>
<img width="1182" alt="Screenshot 2023-05-03 at 16 20 55" src="https://user-images.githubusercontent.com/63515648/236121986-213aa196-b0c5-4132-9b22-bc8654301498.png">
<img width="1438" alt="Screenshot 2023-05-03 at 22 50 22" src="https://user-images.githubusercontent.com/63515648/236121991-5dd9b107-7e61-4b27-8f5c-693dab15cd04.png">
<img width="396" alt="Screenshot 2023-05-03 at 22 49 39" src="https://user-images.githubusercontent.com/63515648/236121995-7f53b3e1-8acd-47b2-821d-c98be9786ff8.png">
<img width="1443" alt="Screenshot 2023-05-03 at 22 49 18" src="https://user-images.githubusercontent.com/63515648/236121996-19b50c60-8663-4930-88e4-aa1c336c9431.png">
<img width="1181" alt="Screenshot 2023-05-03 at 16 21 12" src="https://user-images.githubusercontent.com/63515648/236121999-89bb8727-65b5-46f6-a55a-2ca869e46bca.png">

</details>

# Notes
N/A

# QA Testing Checklist:
Generic Work Type:
- [ ] I can create a Generic Work by filling in the required fields
- [ ] I can create a Generic Work by filling in all the fields

REQ FIELDS DISPLAYED: On the form, I can see the required fields:
- [ ] Creator
- [ ] Rights Statement
- [ ] Date created
- [ ] Resource Type
- [ ] Institution
 
REQ'D FIELDS LABELED/DISPLAYED:
- [ ] On the form, I can see the required fields have a "Required Label" 
- [ ] On the form, I can see the required fields are displayed above the "Additional Fields' button

FORM LABELS: 
- [ ] On the form, I see the correct labels on the create generic work form fields that are in the [spreadsheet](https://docs.google.com/spreadsheets/d/1cyTtAofi2g1sx0ATJuP7ws-77a7OzBXF/edit#gid=1372594593)

HELP TEXT: 
- [ ] On the form, I see the correct help text below the correct field on the create generic work form that are in the [spreadsheet](https://docs.google.com/spreadsheets/d/1cyTtAofi2g1sx0ATJuP7ws-77a7OzBXF/edit#gid=1372594593)

MULTIPLE FIELDS: 
- [ ] On the form, I can add multiple fields that are in the [spreadsheet](https://docs.google.com/spreadsheets/d/1cyTtAofi2g1sx0ATJuP7ws-77a7OzBXF/edit#gid=1372594593)

CONTROLLED VOCABULARY: On the form, I can see the correct controlled vocab show for the following fields:
- [ ] Resource Type - controlled vocab matches list on the [spreadsheet](https://docs.google.com/spreadsheets/d/1cyTtAofi2g1sx0ATJuP7ws-77a7OzBXF/edit#gid=785133840)
- [ ] Institution - controlled vocab matches list on the [spreadsheet](https://docs.google.com/spreadsheets/d/1cyTtAofi2g1sx0ATJuP7ws-77a7OzBXF/edit#gid=785133840)
- [ ] Format - controlled vocab matches list on the [spreadsheet](https://docs.google.com/spreadsheets/d/1cyTtAofi2g1sx0ATJuP7ws-77a7OzBXF/edit#gid=785133840)

ADDITIONAL METADATA DISPLAYED: After clicking the “Additional Fields” button I can see the non-required fields for the GenericWork model:
- [ ] Alternative Title
- [ ] Contributor
- [ ] Abstract
- [ ] Keyword
- [ ] Access Rights
- [ ] Rights notes
- [ ] Publisher
- [ ] Subject
- [ ] Language
- [ ] Identifier
- [ ] Related URL
- [ ] Format
- [ ] Bibliographic citation

EDITING A WORK: When I click edit on a work: 
- [ ] I can see the previously saved field value in the form field
- [ ] I can edit and save the fields and see the updated value present on the saved Generic Work
- [ ] After saving the work and navigating to the work show page, I can see all the metadata listed on the [spreadsheet](https://docs.google.com/spreadsheets/d/1cyTtAofi2g1sx0ATJuP7ws-77a7OzBXF/edit#gid=1372594593)

THE SEACRH RESULTS PAGE: On the search results page you will see the following properties displayed on each work listed in the results.
- [ ] Creator
- [ ] Description (with 300 character limit)
- [ ] Resource Type

THE FACETS THAT SHOULD DISPLAY: On the facets, after creating a generic work with all the form fields filled in. You navigate to the search results page and can see the following facets included:
- [ ] Resource Type
- [ ] Creator
- [ ] Contributor
- [ ] Keyword
- [ ] Subject
- [ ] Language
- [ ] Publisher
- [ ] Date created
- [ ] Institution
- [ ] Format

Please double check the accuracy of the form fields' help text by cross checking from in column G titled "Help Text" of this [spreadsheet](https://docs.google.com/spreadsheets/d/1cyTtAofi2g1sx0ATJuP7ws-77a7OzBXF/edit#gid=1372594593) to the Generic Work form fields we updated listed below:
- [ ] Rights Statement
- [ ] Resource Type
- [ ] Institution
- [ ] Keyword
- [ ] Access Rights
- [ ] Rights notes
- [ ] Publisher
- [ ] Subject
- [ ] Format

Please also check to make sure when you flip through the different languages of those help texts' that the help text appears to change to the alternate language selected for the fields listed above.
- [ ] German
- [ ] Spanish
- [ ] Brazilian Portuguese
- [ ] French
- [ ] Italian
- [ ] Chinese
